compaction/compaction.cc

```diff
@@ -1703,6 +1704,98 @@ future<uint64_t> scrub_validate_mode_validate_reader(flat_mutation_reader_v2 rea
+                        pkey,
+                        scrub_compaction::partition_key_to_string(*last_pkey, schema),
+                        *last_pkey);
+            }
+
+            if (last_pos && pos <= *last_pos) {
+                ++errors;
+                clogger.error("[scrub compaction {}.{}] Out-of-order position {} for partition {} ({}) in index, previous position is {} for partition {} ({})",
+                        schema.ks_name(),
+                        schema.cf_name(),
+                        pos,
+                        scrub_compaction::partition_key_to_string(pkey, schema),
+                        pkey,
+                        *last_pos,
+                        scrub_compaction::partition_key_to_string(*last_pkey, schema),
+                        *last_pkey);
+            }
+
+            auto pr = dht::partition_range::make_singular(dht::decorate_key(schema, pkey));
+            auto reader = sst->make_reader(sst->get_schema(), permit, pr, schema.full_slice(), pc, {}, streamed_mutation::forwarding::no, mutation_reader::forwarding::no);
+            try {
+                auto maybe_ps_opt = co_await reader();
+                if (!maybe_ps_opt) {
+                    ++errors;
+                    clogger.error("[scrub compaction {}.{}] Invalid index entry for partition {} ({}), partition not found in data file",
+                            schema.ks_name(), schema.cf_name(), scrub_compaction::partition_key_to_string(pkey, schema), pkey);
+                } else if (!maybe_ps_opt->is_partition_start()) {
+                    ++errors;
+                    clogger.error("[scrub compaction {}.{}] Invalid index entry for partition {} ({}), entry doesn't point to start of partition",
+                            schema.ks_name(), schema.cf_name(), scrub_compaction::partition_key_to_string(pkey, schema), pkey);
+                } else {
+                    mutation_fragment_v2_opt last_mf;
+                    while (auto next_mf = co_await reader()) {
+                        last_mf = std::move(next_mf);
+                    }
+                    if (!last_mf || !last_mf->is_end_of_partition()) {
+                        ++errors;
+                        clogger.error("[scrub compaction {}.{}] Invalid index entry for partition {} ({}), partition incomplete",
+                                schema.ks_name(), schema.cf_name(), scrub_compaction::partition_key_to_string(pkey, schema), pkey);
+                    }
+                }
+            } catch (...) {
+                ex = std::current_exception();
+            }
+            co_await reader.close();
+            if (ex) {
+                break;
+            }
+
+            last_pkey = std::move(pkey);
+            last_pos = pos;
+            co_await idx_reader.advance_to_next_partition();
+        }
+    } catch (...) {
+        ex = std::current_exception();
+    }
+
+    co_await idx_reader.close();
```

Comment:
We really should figure out a way to integrate these blocking destructors somehow.

Reply:
coroutine::as_future can help somewhat to simplify the code